x-pack/legacy/plugins/infra/public/hooks/use_track_metric.tsx

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useEffect } from 'react';
+import { trackUiMetric } from '../../../../../../src/legacy/core_plugins/ui_metric/public';
+
+/**
+ * Note: The UI Metric plugin will take care of sending this data to the telemetry server.
+ * You can find these metrics stored at:
+ * stack_stats.kibana.plugins.ui_metric.{app}.{metric}(__delayed_{n}ms)?
+ * which will be an array of objects each containing a key, representing the metric, and
+ * a value, which will be a counter
+ */
+
+type ObservabilityApp = 'infra_metrics' | 'infra_logs' | 'apm' | 'uptime';
+
+interface TrackProps {
+  app: ObservabilityApp;
+  delay?: number; // in ms
+  effectDependencies?: unknown[];
+}
+
+type TrackMetricProps = TrackProps & { metric: string };
+
+export function useTrackMetric({
+  app,
+  metric,
+  delay = 0,
+  effectDependencies = [],
+}: TrackMetricProps) {
+  useEffect(() => {
+    let decoratedMetric = metric;
+    if (delay > 0) {
+      decoratedMetric += `__delayed_${delay}ms`;
+    }
+    const id = setTimeout(() => trackUiMetric(app, decoratedMetric), Math.max(delay, 0));
+    return () => clearTimeout(id);
+  }, effectDependencies);
```

Comment:
The fact that none of the closed-over variables are used as a dependency could be seen as "suprising". For example the metric name could change over time as routing takes place, but the effect is not triggered again. Maybe the `metric` and `delay` should be part of the dependencies?

Reply:
My thinking is that if you want those things to control the re-running of the effect, you should explicitly provide them in the effectDependencies array, sot hat we leave this choice up to the caller.